ios/engine/KMEI/KeymanEngine/Classes/HTTPDownloader.swift

```diff
@@ -0,0 +1,133 @@
+//
+//  HTTPDownloader.swift
+//  KeymanEngine
+//
+//  Created by Gabriel Wong on 2017-09-15.
+//  Copyright © 2017 SIL International. All rights reserved.
+//
+
+@objc public protocol HTTPDownloadDelegate: class {
+  func downloadRequestStarted(_ request: HTTPDownloadRequest)
+  func downloadRequestFinished(_ request: HTTPDownloadRequest)
+  func downloadRequestFailed(_ request: HTTPDownloadRequest)
+  func downloadQueueFinished(_ queue: HTTPDownloader)
+}
+
+public class HTTPDownloader: NSObject, URLSessionDelegate, URLSessionTaskDelegate, URLSessionDownloadDelegate,
+URLSessionDataDelegate {
+  var queue: [HTTPDownloadRequest] = []
+  // TODO: Make unowned
+  weak var handler: HTTPDownloadDelegate?
+  var currentRequest: HTTPDownloadRequest?
+  var downloadSession: URLSession!
+  @objc public var userInfo: [String: Any] = [:]
+
+  @objc public init(_ handler: HTTPDownloadDelegate?) {
+    super.init()
+    self.handler = handler
+
+    let config = URLSessionConfiguration.default
+    downloadSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)
+  }
+
+  @objc public func addRequest(_ request: HTTPDownloadRequest) {
+    queue.append(request)
+  }
+
+  func popRequest() -> HTTPDownloadRequest? {
+    if queue.isEmpty {
+      return nil
+    }
+    return queue.remove(at: 0)
+  }
+
+  // Starts the queue.  The queue is managed via event messages in order to process them sequentially.
+  @objc public func run() {
+    if !queue.isEmpty {
+      runRequest()
+    }
+  }
+
+  // Runs the next step in the queue, if appropriate.
+  func runRequest() {
+    // Perhaps more of an 'if', relying on 'completed' messages to continue the loop?
+    if !queue.isEmpty {
+      let req: HTTPDownloadRequest! = popRequest()
+      currentRequest = req
+
+      if req.typeCode == .downloadFile {
+        req.task = downloadSession.downloadTask(with: req.url)
+        handler?.downloadRequestStarted(req)
+      } else if req.typeCode == .downloadCachedData {
+        req.task = downloadSession.dataTask(with: req.url)
+      }
+      req?.task?.resume()
+    } else {
+      handler?.downloadQueueFinished(self)
+    }
+    // The next step in the queue, should it exist, will be triggered by urlSession(_, task:, didCompleteWithError:)
+    // which calls this method.  Thus, this method may serve as the single source of the 'queue finished' message.
+  }
+
+  // This is triggered before the 'didCompleteWithError' delegate method.
+  public func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask,
+                         didFinishDownloadingTo location: URL) {
+    guard let currentRequest = currentRequest else {
+      return
+    }
```

Comment:
This `guard` statement didn't have a correlate in the original Obj-C, and I can't see where the `else` would even trigger - is it basically doing a `nil` check?

Reply:
Yes it's a `nil` check to satisfy the compiler